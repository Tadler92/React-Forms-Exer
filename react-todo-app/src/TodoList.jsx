import { useState } from "react";
import { v4 as uuid } from "uuid";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

const TodoList = () => {
  const INITIAL_STATE = [
    {id: uuid(), taskTodo: 'read'},
    {id: uuid(), taskTodo: 'write'},
    {id: uuid(), taskTodo: 'swim'}
  ];

  const [todos, setTodos] = useState([]);

  const addTodo = (newTodo) => {
    setTodos(todos => [...todos, {...newTodo, id: uuid()}])
  }

  const removeTodo = (id) => {
    setTodos(todos => todos.filter(todo => todo.id !== id));
  }

  return (
    <div>
      <h3>Todo App</h3>
      <NewTodoForm addTodo={addTodo} />
      {todos.map(({id, taskTodo}) => (
        <Todo 
          id={id} 
          taskTodo={taskTodo} 
          key={id} 
          removeTodo={removeTodo}
        />
      ))}
    </div>
  )
}


export default TodoList;