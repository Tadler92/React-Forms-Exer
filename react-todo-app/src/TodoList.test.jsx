import {render, fireEvent, waitFor} from "@testing-library/react";
import TodoList from "./TodoList";
import { expect, it } from "vitest";


describe("TodoList Component", () => {
  it("renders without crashing", function () {
    render(<TodoList />);
  });
  
  
  it("matches snapshot", function () {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should add new todo", function () {
    const {queryByText, getByLabelText} = render(<TodoList />)
    const input = getByLabelText("Task Todo:");
    const btn = queryByText("Add Todo");
    expect(queryByText('Exercise')).not.toBeInTheDocument();
    expect(queryByText('X')).not.toBeInTheDocument();
  
    fireEvent.change(input, {target: {value: 'Exercise'}});
    fireEvent.click(btn);
    waitFor(() => expect(queryByText('Exercise')).toBeInTheDocument());
    waitFor(() => expect(queryByText('X')).toBeInTheDocument());
  });

  it("should remove todo", function () {
    const {queryByText, getByLabelText} = render(<TodoList />)
    const input = getByLabelText("Task Todo:");
    const btn = queryByText("Add Todo");
    let removeBtn = queryByText('X');

    expect(queryByText('homework')).not.toBeInTheDocument();
    expect(removeBtn).not.toBeInTheDocument();
  
    fireEvent.change(input, {target: {value: 'homework'}});
    fireEvent.click(btn);
    waitFor(() => expect(queryByText('homework')).toBeInTheDocument());
    waitFor(() => expect(queryByText('X')).toBeInTheDocument());

    removeBtn = queryByText('X');
    fireEvent.click(removeBtn);
    waitFor(() => expect(queryByText('homework')).not.toBeInTheDocument());
    waitFor(() => expect(removeBtn).not.toBeInTheDocument());
  });
});