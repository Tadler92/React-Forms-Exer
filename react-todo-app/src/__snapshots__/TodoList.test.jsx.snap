// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList Component > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Todo App
    </h3>
    <form
      class="NewTodoForm"
    >
      <label
        for="taskTodo"
      >
        Task Todo: 
      </label>
      <input
        id="taskTodo"
        name="taskTodo"
        placeholder="Task"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
  </div>
</DocumentFragment>
`;
